tools/cbmc/proofs/DNS/DNSgetHostByName_a/DNSgetHostByName_a_harness.c

```diff
@@ -0,0 +1,62 @@
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "queue.h"
+#include "list.h"
+
+/* FreeRTOS+TCP includes. */
+#include "FreeRTOS_IP.h"
+#include "FreeRTOS_DNS.h"
+#include "FreeRTOS_IP_Private.h"
+
+
+/* This assumes that prvParseDNSReply, FreeRTOS_socket, FreeRTOS_GetUDPPayloadBuffer, pvPortMalloc and xTaskResumeAll
+are implemented correctly. The function func mimics the callback function. This proof also assumes that the size of UDP Payload Buffer
+is bounded by MAX_REQ_SIZE and the length of the hostname is bounded by MAX_HOSTNAME_LEN due to performance issues. */
+
+typedef struct xDNS_Callback {
+	TickType_t xRemaningTime;		/* Timeout in ms */
+	FOnDNSEvent pCallbackFunction;	/* Function to be called when the address has been found or when a timeout has beeen reached */
+	TimeOut_t xTimeoutState;
+	void *pvSearchID;
+	struct xLIST_ITEM xListItem;
+	char pcName[1];
+} DNSCallback_t;
+
+uint32_t prvParseDNSReply(uint8_t *pucUDPPayloadBuffer, size_t xBufferLength, TickType_t xIdentifier) {}
+
+Socket_t FreeRTOS_socket(BaseType_t xDomain, BaseType_t xType, BaseType_t xProtocol){
+	Socket_t xCreatedSocket = malloc(sizeof(Socket_t));
+	return xCreatedSocket;
+}
+
+void * FreeRTOS_GetUDPPayloadBuffer(size_t xRequestedSizeBytes, TickType_t xBlockTimeTicks ) {
+	__CPROVER_assume(xRequestedSizeBytes >= 0 && xRequestedSizeBytes <= MAX_REQ_SIZE);
+	void *pvReturn = malloc(MAX_REQ_SIZE);
+	return pvReturn;
+}
```

Comment:
This function doesn't work for several reasons:
* You're assuming that `xRequestedSizeBytes` is between 0 and `MAX_REQ_SIZE`, but you don't actually do anything with `xRequestedSizeBytes` after that.
* More worrying: `MAX_REQ_SIZE` is defined as `100`. So if the code actually calls this function with the value `101` for `xRequestedSizeBytes`, then this becomes `assume(false)`. So that makes the whole proof unsound.

@tautschnig: we were hoping to have a discussion about CBMC warnings at some point, and `assume(false)`-type mistakes do worry me somewhat. I'm not confident that I can catch all of these during review.

I'm not sure, but I think what you want to do here is
```c
void * FreeRTOS_GetUDPPayloadBuffer(size_t xRequestedSizeBytes, TickType_t xBlockTimeTicks ) {
  size_t len;
  __CPROVER_assume(len >= 0 && len <= xRequestedSizeBytes);
  void *pvReturn = malloc(len);
  return pvReturn;
}
```
(I'm not sure whether `len` should be `>=` or `>` 0, if it's the latter then it's of course possible that this function will return a null pointer. Not sure if this is what you want.)

Also, I have a question: is it really possible for the original function to return a buffer that is *smaller* than the requested size? Please confirm this by reading the code, and write a comment above this function explaining why this might happen, e.g. "this function will return a buffer of size up to xRequestedSizeBytes. It may return fewer bytes if the UDP payload was smaller than xRequestedSizeBytes". (I made this up, but please confirm what the original code does).

Reply:
> @tautschnig: we were hoping to have a discussion about CBMC warnings at some point, and assume(false)-type mistakes do worry me somewhat. I'm not confident that I can catch all of these during review.

We should actually include a report of assumptions in CBMC's output. Discussed with others yesterday, I'll try to make a stab at it over the next days.

Reply:
That would be excellent, thank you! But hopefully the assumptions would (also) be machine-readable, or at least in a format that we can grep for. Having them for human consumption only works if somebody regularly reads the output. In a CI environment, the proof should ideally break if something changes that makes an assumption become false.

Reply:
If I make `len <= xRequestedSizeBytes` cbmc is just stuck while running propositional reduction. It generated 3303 VCCs and 2397 were remaining after simplification. It is not able to prove/disprove anything even in 2 hrs.